File: web/tslog/src/iterators.test.ts
import { test, expect } from "bun:test";

// 迭代器与数组方法：map / filter / reduce / zip / flatMap / partition。
// 对应 Rust rustlog/src/iterators.rs 与 Java Stream。
//
// 对应关系速查：
// - map          → .map()
// - filter       → .filter()
// - take(n)      → .slice(0, n)；skip(n) → .slice(n)
// - enumerate    → .entries()
// - zip          → 双数组 map 配对
// - flatten      → .flat() / .flatMap()
// - fold/reduce  → .reduce()
// - partition    → .reduce() 分成两组
// - any / all    → .some() / .every()

test("map / filter / take / skip", () => {
  // map：平方（对应 1..10 平方）
  expect([1, 2, 3, 4, 5, 6, 7, 8, 9, 10].map((x) => x * x).slice(0, 3)).toEqual([1, 4, 9]);

  // filter：长度≤3 的单词
  const words = ["rust", "go", "python", "java", "c++", "js"];
  expect(words.filter((w) => w.length <= 3)).toEqual(["go", "c++", "js"]);

  // take / skip（对应 islice）
  const seq = Array.from({ length: 10 }, (_, i) => i + 1); // 1..10
  expect(seq.slice(0, 3)).toEqual([1, 2, 3]); // take(3)
  expect(seq.slice(7)).toEqual([8, 9, 10]); // skip(7)

  // step_by(5)：按索引步进取元素
  expect(seq.filter((_, i) => i % 5 === 0)).toEqual([1, 6]); // 索引 0、5
  expect(Array.from({ length: 5 }, (_, i) => i * 5)).toEqual([0, 5, 10, 15, 20]); // 对应 0..20 step 5
});

test("enumerate / zip / chain / flatten", () => {
  // enumerate：带索引遍历（对应 Rust enumerate）
  expect(["a", "b", "c"].entries()).toBeInstanceOf(Iterator); // 迭代器对象
  expect([...["a", "b", "c"].entries()]).toEqual([
    [0, "a"],
    [1, "b"],
    [2, "c"],
  ]);

  // zip：双数组一一配对
  const names = ["Alice", "Bob", "Charlie"];
  const scores = [95, 87, 92];
  expect(names.map((n, i) => [n, scores[i]])).toEqual([
    ["Alice", 95],
    ["Bob", 87],
    ["Charlie", 92],
  ]);

  // chain：拼接
  expect([1, 2, 3].concat(10, 11, 12)).toEqual([1, 2, 3, 10, 11, 12]);

  // flatten：展平嵌套数组（对应 flat_map）
  expect([[1, 2], [3, 4, 5], [6]].flat()).toEqual([1, 2, 3, 4, 5, 6]);

  // flat_map：把每个单词展开成字符
  expect(["hello", "world"].flatMap((w) => [...w])).toEqual([
    "h", "e", "l", "l", "o", "w", "o", "r", "l", "d",
  ]);
});

test("filter_map：选出合法数字", () => {
  const strs = ["123", "abc", "456", "not_a_num", "789"];
  const nums = strs.filter((s) => /^\d+$/.test(s)).map(Number);
  expect(nums).toEqual([123, 456, 789]);
});

test("reduce / sum / min / max / any / all", () => {
  const v = [3, 1, 4, 1, 5, 9, 2, 6];

  // reduce 累加（对应 fold 1..10 = 55）
  expect(Array.from({ length: 10 }, (_, i) => i + 1).reduce((acc, x) => acc + x, 0)).toBe(55);

  expect(v.reduce((a, b) => a + b, 0)).toBe(31); // sum
  expect(Math.min(...v)).toBe(1); // min
  expect(Math.max(...v)).toBe(9); // max

  expect(v.some((x) => x > 10)).toBe(false); // any > 10?
  expect(v.every((x) => x > 0)).toBe(true); // all > 0?
});

test("partition：奇偶分组", () => {
  const v = [3, 1, 4, 1, 5, 9, 2, 6];
  const { even, odd } = v.reduce<{ even: number[]; odd: number[] }>(
    (acc, x) => {
      (x % 2 === 0 ? acc.even : acc.odd).push(x);
      return acc;
    },
    { even: [], odd: [] },
  );
  expect(even).toEqual([4, 2, 6]);
  expect(odd).toEqual([3, 1, 1, 5, 9]);
});

// 综合实战：R&D 部门 30 岁以上员工的平均月薪
test("综合实战：平均月薪", () => {
  const staff = [
    { dept: "R&D", age: 28, salary: 30000 },
    { dept: "R&D", age: 35, salary: 45000 },
    { dept: "R&D", age: 42, salary: 60000 },
    { dept: "HR", age: 32, salary: 18000 },
    { dept: "R&D", age: 25, salary: 22000 },
    { dept: "Sale", age: 38, salary: 25000 },
  ];

  const query = staff
    .filter((e) => e.dept === "R&D") // 先筛选部门
    .filter((e) => e.age >= 30) // 再筛选年龄
    .map((e) => e.salary); // 提取月薪

  const avg = query.length === 0 ? 0 : query.reduce((a, b) => a + b, 0) / query.length;
  expect(avg).toBe(52500); // (45000+60000)/2 = 52500
});

File: go/golog/iterators_test.go
package main

import (
	"fmt"
	"iter"
	"slices"
	"strconv"
	"testing"
)

/*
迭代器（Go 1.23+ range over func）：自定义 iter.Seq 迭代器 + 组合操作
对应 Rust rustlog iterators.rs、C# cslog IteratorDemo.cs（LINQ）

Go 1.23 新特性 range over func：
- iter.Seq[V]    单值迭代器   func(yield func(V) bool)
- iter.Seq2[K,V] 双值迭代器   func(yield func(K,V) bool)（对应 map/索引遍历）
- 标准库配套：slices.Collect / slices.Sorted / maps.Keys / maps.Values
*/

// map 变换：对每个元素应用函数（对应 Rust map / C# Select）
func MapSeq[V, R any](seq iter.Seq[V], f func(V) R) iter.Seq[R] {
	return func(yield func(R) bool) {
		for v := range seq {
			if !yield(f(v)) {
				return
			}
		}
	}
}

// filter 过滤：保留满足条件的元素（对应 Rust filter / C# Where）
func FilterSeq[V any](seq iter.Seq[V], keep func(V) bool) iter.Seq[V] {
	return func(yield func(V) bool) {
		for v := range seq {
			if keep(v) && !yield(v) {
				return
			}
		}
	}
}

// take：取前 n 个（对应 Rust take / C# Take）
func TakeSeq[V any](seq iter.Seq[V], n int) iter.Seq[V] {
	return func(yield func(V) bool) {
		i := 0
		for v := range seq {
			if i >= n {
				return
			}
			if !yield(v) {
				return
			}
			i++
		}
	}
}

// drop：跳过前 n 个（对应 Rust skip / C# Skip）
func SkipSeq[V any](seq iter.Seq[V], n int) iter.Seq[V] {
	return func(yield func(V) bool) {
		i := 0
		for v := range seq {
			if i >= n {
				if !yield(v) {
					return
				}
			} else {
				i++
			}
		}
	}
}

// 迭代器：map / filter / take / drop（对应 Rust map/filter、C# Select/Where）
func TestIterMapFilter(t *testing.T) {
	// 1..10 的平方（对应 Rust map / C# Select）
	squares := slices.Collect(MapSeq(slices.Values([]int{1, 2, 3, 4, 5, 6, 7, 8, 9, 10}), func(n int) int {
		return n * n
	}))
	fmt.Println("1..10 平方:", squares) // [1 4 9 16 25 36 49 64 81 100]

	// 过滤长度 ≤ 3 的单词（对应 Rust filter / C# Where）
	words := []string{"go", "c++", "javascript", "rust", "java", "js"}
	shortWords := slices.Collect(FilterSeq(slices.Values(words), func(w string) bool {
		return len(w) <= 3
	}))
	fmt.Println("长度≤3 的单词:", shortWords) // [go c++ js]

	// take / drop（对应 Rust take/skip、C# Take/Skip）
	nums := slices.Values([]int{1, 2, 3, 4, 5, 6, 7, 8, 9, 10})
	fmt.Println("take(3):", slices.Collect(TakeSeq(nums, 3))) // [1 2 3]
	fmt.Println("skip(7):", slices.Collect(SkipSeq(nums, 7))) // [8 9 10]
}

// zip / 展平（对应 Rust zip、C# Zip + SelectMany）
func TestIterZipFlatten(t *testing.T) {
	names := []string{"Alice", "Bob", "Charlie"}
	scores := []int{95, 87, 92}

	// 配对（Seq2 迭代器：yield 返回两个值）
	pairs := slices.Collect(func(yield func(string) bool) {
		for i, name := range names {
			if i >= len(scores) {
				break
			}
			if !yield(fmt.Sprintf("(%s,%d)", name, scores[i])) {
				return
			}
		}
	})
	fmt.Println("zip 配对:", pairs) // [(Alice,95) (Bob,87) (Charlie,92)]

	// 展平嵌套切片（对应 Rust flat_map / C# SelectMany）
	nested := [][]int{{1, 2}, {3, 4}, {5, 6}}
	flattened := slices.Collect(func(yield func(int) bool) {
		for _, arr := range nested {
			for _, v := range arr {
				if !yield(v) {
					return
				}
			}
		}
	})
	fmt.Println("展平:", flattened) // [1 2 3 4 5 6]
}

// reduce / any / all（对应 Rust fold、C# Aggregate/Any/All）
func TestIterReduceAnyAll(t *testing.T) {
	nums := []int{1, 2, 3, 4, 5, 6, 7, 8, 9, 10}

	// reduce 求和（手写循环 + 遍历迭代器）
	sum := 0
	for v := range slices.Values(nums) {
		sum += v
	}
	fmt.Println("sum 1..10 =", sum) // 55

	// any：是否存在 > 10（提前返回）
	anyGreater := false
	for v := range slices.Values(nums) {
		if v > 10 {
			anyGreater = true
			break
		}
	}
	fmt.Println("any > 10?", anyGreater) // false

	// all：是否全部 > 0（出现不满足即返回）
	allPositive := true
	for v := range slices.Values(nums) {
		if v <= 0 {
			allPositive = false
			break
		}
	}
	fmt.Println("all > 0? ", allPositive) // true
}

// 综合实战：R&D 部门 30+ 员工平均月薪（对应 Go/Java/C# 综合示例）
func TestIterAverageSalary(t *testing.T) {
	type Person struct {
		Name   string
		Dept   string
		Age    int
		Salary int
	}
	staff := []Person{
		{Name: "Alice", Dept: "R&D", Age: 28, Salary: 45000},
		{Name: "Bob", Dept: "R&D", Age: 35, Salary: 55000},
		{Name: "Charlie", Dept: "R&D", Age: 32, Salary: 50000},
		{Name: "David", Dept: "HR", Age: 40, Salary: 30000},
		{Name: "Eve", Dept: "R&D", Age: 25, Salary: 35000},
	}

	// 链式组合：过滤 → 提取月薪 → 求平均
	avg := func() int {
		total, count := 0, 0
		for p := range slices.Values(staff) {
			if p.Dept == "R&D" && p.Age >= 30 {
				total += p.Salary
				count++
			}
		}
		if count == 0 {
			return 0
		}
		return total / count
	}()
	fmt.Println("R&D 30+ 员工平均月薪:", avg, "元/月") // 52500
}

// 迭代器组合（filter_map 模式：解析合法数字，对应 Rust filter_map）
func TestIterFilterMap(t *testing.T) {
	maybeNumbers := []string{"123", "abc", "456", "78x", "789"}

	valid := slices.Collect(func(yield func(int) bool) {
		for s := range slices.Values(maybeNumbers) {
			if n, err := strconv.Atoi(s); err == nil {
				if !yield(n) {
					return
				}
			}
		}
	})
	fmt.Println("filter_map 选出合法数字:", valid) // [123 456 789]
}
